src/main/scala/com/microsoft/hyperspace/Hyperspace.scala

```diff
@@ -166,18 +171,30 @@ class Hyperspace(spark: SparkSession) {
 }
 
 object Hyperspace {
-  private lazy val context = new ThreadLocal[HyperspaceContext]
+  // Maintain HyperspaceContext for each Spark session, since Hyperspace depends on
+  // the session's properties such as configs, etc.
+  // Value as (thread id, hyperspace context) to enforce one client thread per Spark session
+  // as Hyperspace doesn't guarantee thread safety.
+  private[hyperspace] lazy val contexts =
+    new ConcurrentHashMap[SparkSession, (Long, HyperspaceContext)]().asScala
 
   private[hyperspace] def getContext(spark: SparkSession): HyperspaceContext = {
-    val ctx = context.get()
-    if (ctx == null || !ctx.spark.equals(spark)) {
-      // Ensure the Spark session the caller provides is the same as
-      // the one HyperspaceContext is using because Hyperspace depends on the
-      // session's properties such as configs, etc.
-      context.set(new HyperspaceContext(spark))
+    spark.sparkContext.isStopped
+    val ctx = contexts.get(spark)
+    if (!ctx.isDefined) {
+      contexts.synchronized {
+        if (contexts.size >= HyperspaceConf.maxSupportedSparkSessions(spark)) {
+          throw HyperspaceException("Too many Spark sessions uses Hyperspace." +
+            "Please increase the config: spark.hyperspace.maxSessions")
+        }
+        contexts.put(spark, (Thread.currentThread().getId, new HyperspaceContext(spark)))
+      }
+    } else if (ctx.get._1 != Thread.currentThread().getId) {
+      throw HyperspaceException(s"Hyperspace does not support multiple threads " +
```

Comment:
@imback82 
I guess this is too restricted.. How about writing some warning log and creating a new context in this case?